### PR TITLE
WIP Kflag example

### PR DESCRIPTION
**Pull request by mtaufen**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Example implementation of a wrapper library around pflag as proposed in this KEP: https://github.com/kubernetes/enhancements/pull/764

```release-note
NONE
```

/hold
/wg component-standard
/sig api-machinery
/sig cluster-lifecycle

### CODE REVIEW

File: staging/src/k8s.io/component-base/kflag/kflag.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	v := &Int32Value{
+		name: name,
+		fs:   fs.fs,
+	}
+	fs.fs.Int32Var(&v.value, name, def, usage)
+	return v
+}
+
+// Replace the int32 altogether
+func (v *Int32Value) Replace(target *int32) {
```
Comment (sttts) [inline staging/src/k8s.io/component-base/kflag/kflag.go]: `Replace` is a strange name. `ApplyIfChanged` ?

File: staging/src/k8s.io/component-base/kflag/kflag.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+	if v.fs.Changed(v.name) {
+		*target = v.value
+	}
+}
+
+// Custom user-defined behavior
+func (v *Int32Value) Custom(apply func(value int32)) {
```
Comment (sttts) [inline staging/src/k8s.io/component-base/kflag/kflag.go]: method names should be verbs. What is the verb here?

File: staging/src/k8s.io/component-base/kflag/kflag.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	}
+	for k, v := range def {
+		val.value[k] = v
+	}
+	fs.fs.Var(utilflag.NewMapStringBool(&val.value), name, usage)
+	return val
+}
+
+// Replace the map altogether
+func (v *MapStringBoolValue) Replace(target *map[string]bool) {
```
Comment (sttts) [inline staging/src/k8s.io/component-base/kflag/kflag.go]: I suggested `ApplyIfChanged` above, but obviously here `Replace`-vs-`Merge` has some sense. Still `Replace` is somehow strange for values. Actually it is a setter, no? Maybe just `Set`?

File: staging/src/k8s.io/component-base/kflag/example/options/options.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	c := &config.Config{}
+	config.Default(c)
+	return c
+}
+
+// We could have AddConfigFlags just return the function to apply
+// config flags, to keep each flag's registration and application
+// code close together:
+
+func AddConfigFlags(fs *kflag.FlagSet) (apply func(c *config.Config)) {
```
Comment (rosti) [inline staging/src/k8s.io/component-base/kflag/example/options/options.go]: Probably have a helper based on this inside the library itself? I can see many k8s components having to implement sort of the same thing here.

File: staging/src/k8s.io/component-base/kflag/example/options/options.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	OtherNum int32
+}
+
+func NewFlags() *Flags {
+	return &Flags{
+		OtherNum: 4,
+	}
+}
+
+func NewConfig() *config.Config {
```
[the remark this answers is not in the document]
Comment (mtaufen) [reply staging/src/k8s.io/component-base/kflag/example/options/options.go]: Mostly by Kubelet's historical convention, this is where the constructor is: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/options/options.go#L291. 

A few probable reasons it's stayed this way:
1. The API machinery doesn't have very convenient helpers to produce a defaulted internal type for a given API version, so in the Kubelet that "get scheme, get versioned, default versioned, convent versioned to internal" sequence happens here.
2. In the Kubelet we have different defaults for flags vs config files, and those legacy defaults are set up by the constructor in the options package (which deals with flags).

File: staging/src/k8s.io/component-base/kflag/kflag.go (hunk 2 above)
Comment (mtaufen) [reply staging/src/k8s.io/component-base/kflag/kflag.go]: Maybe we could call it `CustomApply`, or just `Apply`.

File: staging/src/k8s.io/component-base/kflag/kflag.go (hunk 1 above)
Comment (mtaufen) [reply staging/src/k8s.io/component-base/kflag/kflag.go]: I thought about just calling this `Set`, but given that there is a `Replace` vs. `Merge` distinction for other types, and `Replace` and `Set` basically mean the same thing, I figured we could keep a consistent naming convention and call it `Replace` for all types. Alternatively, we could just call it `Set` for all types, which might be clearer (as you mentioned below).

I'd like to leave off the `IfChanged`, since the plan is for all apply methods to be conditional on the flag being set.

File: staging/src/k8s.io/component-base/kflag/kflag.go (hunk 3 above)
Comment (mtaufen) [reply staging/src/k8s.io/component-base/kflag/kflag.go]: I'd be ok with `Set`

File: staging/src/k8s.io/component-base/kflag/example/options/options.go (hunk 4 above)
Comment (mtaufen) [reply staging/src/k8s.io/component-base/kflag/example/options/options.go]: Since this helper needs to understand the `Config` type, which kflag can't know about (it is different for any given component), I don't think we would be able to maintain compile-time type safety if we did that.

**Review by thockin — COMMENTED** [review]
I get where this is going, but it feels like we could go even further?  Why not erase the distinction between Flags and ConfigFlags?  At the end of the day there is a single config structure that the app wants to know about, right?

File: staging/src/k8s.io/component-base/kflag/kflag.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/component-base/kflag/kflag.go]: both are fine, maybe `Apply` is best

File: staging/src/k8s.io/component-base/kflag/kflag.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/component-base/kflag/kflag.go]: `Set`, `Merge`, `Apply` – looks like a good triple of verbs.

File: staging/src/k8s.io/component-base/kflag/example/options/options.go (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/component-base/kflag/example/options/options.go]: It will be some boilerplate code to be implemented. But I think it is simple enough to be not an issue. And having type safety forces us to do that again and again unfortunately.

File: staging/src/k8s.io/component-base/kflag/example/example.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+	"github.com/spf13/pflag"
+
+	"k8s.io/component-base/kflag"
+	"k8s.io/component-base/kflag/example/config"
+	"k8s.io/component-base/kflag/example/options"
+)
+
+// go run example.go --num=3 --other-num=6 --map=enableFoo=false,enableBaz=true
+func main() {
+	fs := kflag.NewFlagSet(pflag.NewFlagSet("component", pflag.ExitOnError))
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/component-base/kflag/example/example.go]: :+1:

**Comment by sttts** [comment]
> I get where this is going, but it feels like we could go even further? Why not erase the distinction between Flags and ConfigFlags? At the end of the day there is a single config structure that the app wants to know about, right?

It's not that simple as "a single config structure".

We have a distinction between flags and configs because they do not map 1:1. In the apiservers actually we call `Flags` options and then get this pattern:

- `ComponentConfig` – is the internal config type visible to the user. It's an API type, without any logic attached.
- `Options` (called `Flags` here) – subsumes the `ComponentConfig`, but might have more setting usefull only for plumbing the component, but not user-facing (example: `k8s.io/apiserver` has a `RecommendedOptions` struct which has a field `RecommendedPluginOrder`, later showing up as default value for the admission plugins. This is not user-facing, but purely for plumbing. Many more similar fields exist.
- server `Config` – is the the actual server runtime config, might include clients, informers, certs, ... Actually in some places we call this `Server` instead of `Config`.

In short:
- `Options = ComponentConfig + plumbing fields + flags not existing in ComponentConfig`
- `Config` is derived from `Options` by interpreting the user intent and by creating runtime objects.

Hence, we cannot collapse `Flags` and `Config`.

**Comment by thockin** [comment]
@sttts I get that there are some distinctions.  I am saying that, from the POV of "I just want to use this to do the right thing" this is complicated and presents a distinction that doesn't seem very meaningful to me.  Can we absorb more of the complexity so components follow simpler patterns?

File: staging/src/k8s.io/component-base/kflag/example/options/options.go (hunk 4 above)
Comment (mtaufen) [reply staging/src/k8s.io/component-base/kflag/example/options/options.go]: Don't we all ;)

**Comment by mtaufen** [comment]
This PR is much more about making it easier to map flags to arbitrary structures, than it is about how config is structured and plumbed internally after parsing out of flags, and I'd like to avoid jumping down that rabbit-hole just yet.

**Comment by thockin** [comment]
establishing a common base that is as simple as possible for people to copy
correctly seems pretty important, though?

Put another way: you make this change, 10 people copy it.  Then you make it
better - what impetus is there for everyone keeping pace?

**Comment by mtaufen** [comment]
> establishing a common base that is as simple as possible for people to copy correctly seems pretty important, though?

Of course. But it's not going to materialize all at once.

Continuous adoption is going to be easier in some areas than others. Flag registration, parsing, and application is relatively isolated to the bootstrap sequences of components. Config structure (even just the internal, non-user-facing stuff) threads through the entire machine, and is much more tedious to change. User-facing API conventions even more so.

If you want a componentconfig API but don't have it yet, there are fewer backwards-compatibility tasks/headaches in your future after kflag than before. Given that nobody really wants to copy the complex flag-precedence implementation in the Kubelet, and this pretty much blocks adopting componentconfig, I'd say this part is rather important in and of itself.

> Then you make it better - what impetus is there for everyone keeping pace?

"Better" should mean I'll have less future work to do, or fewer future headaches to deal with, or that I get some other value disproportionate to the work involved to stay up to date. I'd expect the same criteria to apply (both as qualification for "better," and incentive to change) for any other standard we come up with.

**Comment by mtaufen** [comment]
I spent some time playing around with template-based codegen today.
Looks like it will probably work to generate helpers for the basic types supported by pflag.

Maps other than map[string]string are trickier, because we need string to sometype value parsers, and those methods are harder to template (e.g. map[string]bool needs strconv.ParseBool).

Haven't tried slice types yet, which may be easier than maps given how many are implemented by pflag.

Note that pflag appears to have some map types, but K8s mostly uses custom map parsers with different features/behavior.
